apps/emqx/src/emqx_channel.erl

```diff
@@ -1137,16 +1137,14 @@ handle_call(
     kick,
     Channel = #channel{
         conn_state = ConnState,
-        will_msg = WillMsg,
-        clientinfo = ClientInfo,
         conninfo = #{proto_ver := ProtoVer}
     }
 ) ->
-    (WillMsg =/= undefined) andalso publish_will_msg(ClientInfo, WillMsg),
+    Channel0 = maybe_publish_will_msg(kick, Channel),
```

Comment:
Q: Probably out-of-scope of this PR, but still. If the Will message has a delay, in case of MQTTv5 we'll arm the timer in `maybe_publish_will_msg/2` but then just shut down the connection process right away, right?

Reply:
For MQTTv5,
If the session is not timeout, it won't shutdown connection process right away.

If session timeout first, then connection process shutdown will lead to willmsg publish with reason ({shutdown, expired})
if willmsg timeout first when the connection process is still alive, connection process will publish the willmsg and set willmsg to null.

Such cases are covered in paho python test.

Reply:
I probably need to clarify that the question is specifically about `handle_call(kick, ...)`, which is probably not part of the paho test suite.

So imagine that the Will Message has 5 seconds delay. What will happen here AFAICT:
1. Call to `maybe_publish_will_msg/2` will arm will message timer to fire 5 seconds after.
2. Call to `shutdown(kicked, ok, ...)` will make `handle_call/2` to return with `{shutdown, kicked, ok, ...}`.
3. Outer `emqx_connection:handle_call/3` will translate that to `{stop, {shutdown, kicked}, ...}`.
4. This will effectively translate to `emqx_channel:terminate({shutdown, kicked}, ...)` call.
5. This clause in `terminate/2` does nothing with will messages, so the process will just exit without sending the Will Message.

Session Expiry Interval doesn't seem to affect anything in this codepath.

Reply:
right.

 here it does not even check if the connection is allowed. (passing REAUTH ??) and kick means closing connection and deleting session (session ends).  according to the spec:

```
The Server SHOULD publish Will Messages promptly after the Network Connection is closed and the Will Delay Interval has passed, or when the Session ends, whichever occurs first. 
```

shall I just ignore 'promptly' and publish it in `terminate/2` after send 'DISCONNECT' to client?  hum... by sending 'DISCONNECT' the conn process here may stuck for a while.

I will see if I could handle it in maybe_publish_will_msg by checking the conn_state and the reason....

Reply:
I think "Session ends" here means EMQX connection process terminates due to `{shutdown, expired}`( session expired) OR `{shutdown, kicked}` (session deleted).

But I don't know if there are more 'reasons' casues  "Session ends".